lib/pacemaker/pcmk_sched_colocation.c

```diff
@@ -1439,26 +1437,41 @@ pcmk__apply_coloc_to_priority(pcmk_resource_t *dependent,
         return;
     }
 
+    if (colocation->primary_role != pcmk_role_unknown) {
+        /* Colocation applies only if the primary's next role matches
+         *
+         * @TODO Why ignore a mandatory colocation in this case when we apply
+         * its negation in the mismatched value case?
+         */
```

Comment:
Let's say we have mandatory "promoted dependent with promoted primary". If the next role of the primary instance that we're considering is NOT promoted, we simply ignore the colocation.

My thought was that we should prevent the dependent from promoting with that primary instance.

Otherwise, suppose none of the primary instances' next role is promoted. Then we'll end up ignoring this mandatory colocation and promoting the dependent somewhere. That seems problematic.

---

Another, conflicting possible edge case just now occurred to me. Suppose we have `clone-node-max=2` and `promoted-node-max=1`. Suppose `node1` has a promoted primary instance and an unpromoted primary instance. We wouldn't want to ban the dependent from `node1` upon finding the unpromoted primary there.

The same objection might apply to my fix in a later commit in this PR, where if `primary_node == NULL` we ban the dependent from promoting on `dependent_node`. The new tests all use primitive primaries. If the primary is a clone and one of its instances is unassigned, I wonder if that would affect dependent promotion.

Reply:
By the way, if any of my concerns are valid, these might not be NEW problems.